torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -0,0 +1,406 @@
+  for (int i = 0; i < srcSizes.size(); i++) {
+    entry->src[i] = at::zeros(*key.type, at::IntList(srcSizes[i]));
+  }
+
+  return std::move(entry);
+}
+
+EntryType ProcessGroupGloo::checkout(const AlgorithmKey& key) {
+  std::unique_lock<std::mutex> lock(m_);
+
+  // Initialize number of entries for this key.
+  if (cacheCreated_.count(key) == 0) {
+    cacheCreated_.emplace(key, 0);
+  }
+
+  // If there is no entry for this key yet, it must be the first time
+  // we see and can create a new entry. Use hard limit of 1 instance
+  // per key until we add support for a dynamic limit.
+  if (cacheCreated_[key] < 1) {
+    cacheCreated_[key]++;
+    return std::move(construct(key));
+  }
+
+  // Optionally wait for entry to be returned to the cache.
+  auto it = cache_.find(key);
+  while (it == cache_.end()) {
+    cacheCV_.wait(lock);
+    it = cache_.find(key);
+  }
+
+  // Grab entry from the cache and return it.
+  auto entry = std::move(it->second);
+  cache_.erase(key);
+  return std::move(entry);
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupGloo::enqueue(EntryType entry) {
+  auto work = std::make_shared<WorkGloo>();
+  std::unique_lock<std::mutex> lock(m_);
+  queue_.push_back(std::make_tuple(std::move(entry), work));
+  queueProduceCV_.notify_one();
+  return std::move(work);
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupGloo::broadcast(
+    std::vector<at::Tensor>& tensors,
+    const BroadcastOptions& opts) {
+  assertSameSizeAndType(tensors);
+
+  AlgorithmKey key;
+  key.collectiveType = CollectiveType::BROADCAST;
+  key.type = &tensors[0].type();
+  key.devices = getDevices(tensors);
+  key.srcSizes = getSizes(tensors);
+  key.srcRank = opts.rootRank;
+  key.srcTensor = opts.rootTensor;
+
+  // Retrieve (create or wait for) cache entry
+  auto entry = checkout(key);
+
+  // Only copy root tensor
+  if (getRank() == opts.rootRank) {
+    entry->src[opts.rootTensor].copy_(tensors[opts.rootTensor]);
+  }
+
+  // Define how to run the algorithm and copy back results
+  entry->run = [tensors](EntryType& entry) mutable {
+    entry->algorithm->run();
+    for (int i = 0; i < tensors.size(); i++) {
+      tensors[i].copy_(entry->src[i]);
+    }
+  };
+
+  return enqueue(std::move(entry));
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupGloo::allreduce(
+    std::vector<at::Tensor>& tensors,
+    const AllreduceOptions& opts) {
+  assertSameSizeAndType(tensors);
+
+  AlgorithmKey key;
+  key.collectiveType = CollectiveType::ALLREDUCE;
+  key.type = &tensors[0].type();
+  key.srcSizes = getSizes(tensors);
+  key.devices = getDevices(tensors);
+  key.reduceOp = opts.reduceOp;
+
+  // Retrieve (create or wait for) cache entry
+  auto entry = checkout(key);
+
+  // Copy input tensors
+  for (int i = 0; i < tensors.size(); i++) {
+    entry->src[i].copy_(tensors[i]);
+  }
+
+  // Define how to run the algorithm and copy back results
+  entry->run = [tensors](EntryType& entry) mutable {
```

Comment:
`mutable` is unfortunately a C++14 thing. It doesn't look like you need it though. (Same applies above)

Reply:
If I don't specify it I get an error saying the tensors variable is const so I can't call `copy_`. I tried making a copy of the tensors vector and passing that in (as opposed to a ref which I assumed would be copied anyway), but this didn't work either. Any ideas?

Reply:
http://en.cppreference.com/w/cpp/language/lambda says the mutable keyword is C++11 btw.

Reply:
Huh you're right. Idk why I thought it's C++14.